File: src/services/tauri/audio.ts
import { TauriCommand } from './commands';
import type { TauriCommandArgs, TauriCommandResult } from './contracts';
import { invokeTauri } from './invoke';

export type AudioDevice =
  TauriCommandResult<typeof TauriCommand.audio.getSystemAudioDevices>[number];

export type StartAudioCaptureRequest =
  TauriCommandArgs<typeof TauriCommand.audio.startSystemAudioCapture>;

export type SetCapturePausedRequest =
  TauriCommandArgs<typeof TauriCommand.audio.setSystemAudioCapturePaused>;

export async function setSystemAudioMute(mute: boolean): Promise<void> {
  await invokeTauri(TauriCommand.audio.setSystemAudioMute, { mute });
}

export async function getSystemAudioDevices(): Promise<AudioDevice[]> {
  return invokeTauri(TauriCommand.audio.getSystemAudioDevices);
}

export async function startSystemAudioCapture(request: StartAudioCaptureRequest): Promise<void> {
  await invokeTauri(TauriCommand.audio.startSystemAudioCapture, request);
}

export async function stopSystemAudioCapture(instanceId: string): Promise<string> {
  return invokeTauri(TauriCommand.audio.stopSystemAudioCapture, { instanceId });
}

export async function setSystemAudioCapturePaused(request: SetCapturePausedRequest): Promise<void> {
  await invokeTauri(TauriCommand.audio.setSystemAudioCapturePaused, request);
}

export async function setMicrophoneBoost(boost: number): Promise<void> {
  await invokeTauri(TauriCommand.audio.setMicrophoneBoost, { boost });
}

export async function getMicrophoneDevices(): Promise<AudioDevice[]> {
  return invokeTauri(TauriCommand.audio.getMicrophoneDevices);
}

export async function startMicrophoneCapture(request: StartAudioCaptureRequest): Promise<void> {
  await invokeTauri(TauriCommand.audio.startMicrophoneCapture, request);
}

export async function stopMicrophoneCapture(instanceId: string): Promise<string> {
  return invokeTauri(TauriCommand.audio.stopMicrophoneCapture, { instanceId });
}

export async function setMicrophoneCapturePaused(request: SetCapturePausedRequest): Promise<void> {
  await invokeTauri(TauriCommand.audio.setMicrophoneCapturePaused, request);
}

File: platforms/desktop/src/commands/audio.rs
use crate::integrations::audio::{AudioDevice, AudioState};
use tauri::{AppHandle, State, Window};

#[tauri::command(async)]
pub fn get_system_audio_devices() -> Result<Vec<AudioDevice>, String> {
    crate::integrations::audio::get_system_audio_devices()
}

#[tauri::command(async)]
pub fn get_microphone_devices() -> Result<Vec<AudioDevice>, String> {
    crate::integrations::audio::get_microphone_devices()
}

#[tauri::command(async)]
pub fn start_system_audio_capture(
    app: AppHandle,
    window: Window,
    state: State<'_, AudioState>,
    sherpa_state: State<'_, crate::integrations::asr::AsrState>,
    device_name: Option<String>,
    instance_id: String,
    output_path: Option<String>,
) -> Result<(), String> {
    crate::integrations::audio::start_system_audio_capture(
        app,
        window,
        state,
        sherpa_state,
        device_name,
        instance_id,
        output_path,
    )
}

#[tauri::command(async)]
pub fn start_microphone_capture(
    app: AppHandle,
    window: Window,
    state: State<'_, AudioState>,
    sherpa_state: State<'_, crate::integrations::asr::AsrState>,
    device_name: Option<String>,
    instance_id: String,
    output_path: Option<String>,
) -> Result<(), String> {
    crate::integrations::audio::start_microphone_capture(
        app,
        window,
        state,
        sherpa_state,
        device_name,
        instance_id,
        output_path,
    )
}

#[tauri::command]
pub async fn stop_system_audio_capture(
    state: State<'_, AudioState>,
    instance_id: String,
) -> Result<String, String> {
    crate::integrations::audio::stop_system_audio_capture(state, instance_id).await
}

#[tauri::command]
pub async fn stop_microphone_capture(
    state: State<'_, AudioState>,
    instance_id: String,
) -> Result<String, String> {
    crate::integrations::audio::stop_microphone_capture(state, instance_id).await
}

#[tauri::command]
pub fn set_system_audio_capture_paused(
    state: State<'_, AudioState>,
    instance_id: String,
    paused: bool,
) -> Result<(), String> {
    crate::integrations::audio::set_system_audio_capture_paused(state, instance_id, paused)
}

#[tauri::command]
pub fn set_microphone_capture_paused(
    state: State<'_, AudioState>,
    instance_id: String,
    paused: bool,
) -> Result<(), String> {
    crate::integrations::audio::set_microphone_capture_paused(state, instance_id, paused)
}

#[tauri::command]
pub fn set_microphone_boost(state: State<'_, AudioState>, boost: f32) -> Result<(), String> {
    crate::integrations::audio::set_microphone_boost(state, boost)
}

#[tauri::command]
pub async fn set_system_audio_mute(mute: bool) -> Result<(), String> {
    crate::platform::system_audio::set_system_audio_mute(mute).await
}
